### PR TITLE
DSG-2893: view attachments in secure messages

### PR DESCRIPTION
**Pull request by lucywang-nv**
# Basics

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Styling and layout changes
- [ ] Code style changes (formatting, local variables)
- [ ] Refactoring (no functional changes)
- [ ] Build or CI related changes
- [ ] Tests
- [ ] Other... Please describe:

## What changes does this PR introduce?
Allow users to be able to view attachments in secure messages.

## Screenshot of the affected changes
https://www.loom.com/share/64b7caabfc6c40898b588dbba2d03bca

Latest commit removes restriction on width, unlike in video:
image

## Checklist

- [ ] I have updated the documentation accordingly
- [ ] I have confirmed all acceptance criteria defined by the story have been met
- [ ] I have confirmed that Sonar Qube findings for new code are reviewed and fixed, or will be whitelisted

### CODE REVIEW

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html [hunk 1]
```diff
@@ -49,6 +50,22 @@ <h4 class="message__sender">{{message.senderDisplayName}}</h4>
     </div>
     <ng-template #richText>
       <nuverial-rich-text-viewer [content]="message.body" />
+      <div class="message__view-attachments" *ngIf="message.attachments.length > 0">
```
Comment (asultan001) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.html:53]: @lucywang-nv we may want to encapsulate this logic in a component as we likely plan to use this to support displaying attachments in notes. Alternatively we can refactor this at that point in time

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.spec.ts [hunk 2]
```diff
@@ -164,5 +168,27 @@ describe('ConversationComponent', () => {
       expect(replySpy).not.toHaveBeenCalled();
       expect(replies.length).toBeFalsy();
     });
+
```
Comment (asultan001) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.spec.ts:171]: We should add tests to make sure attachments are displayed accordingly as well as a test to make sure the attachment indicator is displayed when there are attachments to show in a message.

File: libs/shared/feature/documents/src/lib/services/document-form.service.ts [hunk 3]
```diff
@@ -15,6 +15,22 @@ import { Observable, last, retry, switchMap, takeWhile, tap, timer } from 'rxjs'
...
 
+  public downloadDocument$(documentId: string, documentName: string): Observable<Blob> {
+    return this._documentApiService.getDocumentFileData$(documentId).pipe(
+      tap(data => {
+        const fileURL = window.URL.createObjectURL(data);
+        const link = document.createElement('a');
+        link.href = fileURL;
+        link.download = documentName;
+
+        document.body.appendChild(link);
```
Comment (davidyoung77) [inline libs/shared/feature/documents/src/lib/services/document-form.service.ts:26]: It is best practice in Angular to use Renderer2 to add and remove things from the DOM in this way. LMK if you need guidance on this. we have some examples in the code already

File: libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts [hunk 4]
```diff
@@ -128,6 +140,23 @@ export class ConversationComponent {
...
+            message.attachments.forEach(attachment => {
+              this._documentFormService
+                .getDocumentById$(attachment)
+                .pipe(
+                  take(1),
+                  tap(document => {
+                    this.attachmentNames[attachment] = document.filename;
+                    this.attachmentNamesSize = Object.keys(this.attachmentNames).length;
+                  }),
+                  tap(_ => this._cdr.detectChanges()),
```
Comment (davidyoung77) [inline libs/shared/feature/messaging/src/lib/components/conversation/conversation.component.ts:154]: These can be in the same tap

**Review by davidyoung77 — APPROVED** [review]

File: libs/shared/feature/documents/src/lib/services/document-form.service.ts (hunk 3 above)
Comment (lucywang-nv) [reply libs/shared/feature/documents/src/lib/services/document-form.service.ts]: Just changed.

**Review by davidyoung77 — APPROVED** [review]

**Pull request merged**